### ISSUE TITLE
Terraform fail on EBS resizing

### ISSUE THREAD

**Issue by Seraf** [hashicorp/terraform#12496]
Hello everyone,

I tried to test the new resize feature of EBS volumes. I saw my volume going to a yellow state : "in-use - optimizing (99%)" on the console (I resized from 5go to 10go, the volume was attached), but Terraform failed.

### Terraform Version
Terraform v0.8.8

### Affected Resource(s)
- aws_ebs_volume

### Terraform Configuration Files
```hcl
# Create an EBS to contains "flux" directory used by FTP
resource "aws_ebs_volume" "test" {
  depends_on        = ["aws_instance.test"]
  availability_zone = "${aws_instance.test.availability_zone}"
  type              = "gp2"
  size              = "10"

  tags {
    Name = "test-terraform"
  }
}

resource "aws_volume_attachment" "test" {
  depends_on  = ["aws_ebs_volume.test"]
  device_name = "/dev/xvdg"
  volume_id   = "${aws_ebs_volume.test.id}"
  instance_id = "${aws_instance.test.id}"
}
```

### Debug Output
```
data.aws_ami.debian_jessie_latest: Refreshing state...
aws_instance.test: Refreshing state... (ID: i-06bb2c1af28780e8b)
aws_ebs_volume.test: Refreshing state... (ID: vol-05974ea0922dd518b)
aws_volume_attachment.test: Refreshing state... (ID: vai-1245252251)
aws_ebs_volume.test: Modifying...
  size: "5" => "10"
aws_ebs_volume.test: Still modifying... (10s elapsed)
Error applying plan:

1 error(s) occurred:

* aws_ebs_volume.test: Error waiting for Volume (vol-05974ea0922dd518b) to become available: unexpected state 'in-use', wanted target 'available'. last error: %!s(<nil>)
```
### Expected Behavior
The disk should have been modified and terraform shouldn't fail.

### Actual Behavior
Terraform don't wait when the state is "in-use - optimizing (99%)" and fail.
The operation is still working on AWS side, and finish successfully with the result : "in-use - completed (100%)". But terraform failed.

### Steps to Reproduce
I don't know if it's related to a slow operation on AWS side. It seems that AWS introduced new states since their modification on EBS resizing. Maybe terraform should catch/wait for these ones

Thanks !

**Comment by stack72** [comment]
Hi @Seraf 

I am looking into this right now

Thanks

Paul

**Comment by stack72** [comment]
HI @Seraf 

So i have just tried this, both locally and in the console. The issue you are seeing is that EBS volumes cannot be resized while attached. Terraform gave me the following:

```
% terraform plan                                                                                                             ✹ ✭
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.

aws_instance.test: Refreshing state... (ID: i-0b695636585f042b9)
aws_ebs_volume.test: Refreshing state... (ID: vol-0f493fe770d8c4c1d)
aws_volume_attachment.test: Refreshing state... (ID: vai-1847762407)

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed. Cyan entries are data sources to be read.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

~ aws_ebs_volume.test
    size: "10" => "12"


Plan: 0 to add, 1 to change, 0 to destroy.
```

Then it proceeded to do the following:

```
% terraform apply                                                                                                            ✹ ✭
aws_instance.test: Refreshing state... (ID: i-0b695636585f042b9)
aws_ebs_volume.test: Refreshing state... (ID: vol-0f493fe770d8c4c1d)
aws_volume_attachment.test: Refreshing state... (ID: vai-1847762407)
aws_ebs_volume.test: Modifying...
  size: "10" => "12"
Error applying plan:

1 error(s) occurred:

* aws_ebs_volume.test: IncorrectState: Modification is not supported for volume vol-0f493fe770d8c4c1d in its current state in-use. Please detach the volume or stop the attached instance and retry.
	status code: 400, request id: 280b1ab0-4901-4f72-a349-08eeafca3ec0

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

I can see the same error in the AWS console:

<img width="677" alt="screen shot 2017-03-07 at 17 23 37" src="https://cloud.githubusercontent.com/assets/227823/23663359/08d703bc-035b-11e7-8631-2b6c2d6a658b.png">

I am going to discuss this with some of the team and see the thoughts they have on it

Paul

**Comment by Seraf** [comment]
Hello,

It may depends of the kind of volume, but I just tried it on a "in-use" gp2 volume on the console, and the status become "in-use Modifying (0%)", goes to "in-use - optimizing (99%)", and is now : "in-use - optimizing (100%)"

I didn't have warning/error on the AWS console, so I expect to not have error on the terraform side too.
I saw there were some incompatibilities on standard volumes (magnetic), so there is some conditions to resize an EBS.

Anyway, when I launched the apply, it triggered the modification on my EBS, but the status wasn't "in-use" or "attached" or "detached", but "in-use - optimizing (99%)" and it seems that Terraform was not expecting that.

**Comment by stack72** [comment]
got it - will keep looking into it - what size instance?

**Comment by Seraf** [comment]
Here is the full snippet I'm using :

```
data "aws_ami" "debian_jessie_latest" {
  most_recent = true

  filter {
    name   = "name"
    values = ["debian-jessie-*"]
  }

  filter {
    name   = "virtualization-type"
    values = ["hvm"]
  }

  filter {
    name   = "architecture"
    values = ["x86_64"]
  }

  filter {
    name   = "root-device-type"
    values = ["ebs"]
  }

  owners = ["379101102735"] # Debian
}

resource "aws_instance" "test" {
  ami                         = "${data.aws_ami.debian_jessie_latest.id}"
  associate_public_ip_address = true
  count                       = 1
  instance_type               = "t2.medium"

  root_block_device {
    volume_size           = "10"
    volume_type           = "standard"
    delete_on_termination = true
  }

  tags {
    Name    = "test-terraform"
  }
}

resource "aws_ebs_volume" "test" {
  depends_on        = ["aws_instance.test"]
  availability_zone = "${aws_instance.test.availability_zone}"
  type       = "gp2"
  size              = "10"

  tags {
    Name = "test-terraform"
  }
}

resource "aws_volume_attachment" "test" {
  depends_on  = ["aws_ebs_volume.test"]
  device_name = "/dev/xvdg"
  volume_id   = "${aws_ebs_volume.test.id}"
  instance_id = "${aws_instance.test.id}"
}
```
I'm working on eu-west-1 region

**Comment by kshcherban** [comment]
Reproduced with v0.9.2.
```
module.ebs_builder.aws_ebs_volume.ebs_data_volume: Still modifying... (ID: vol-027e83f7, 10s elapsed)
Error applying plan:

1 error(s) occurred:

* module.ebs_builder.aws_ebs_volume.ebs_data_volume: 1 error(s) occurred:

* aws_ebs_volume.ebs_data_volume: Error waiting for Volume (vol-027e83f7) to become available: unexpected state 'in-use', wanted target 'available'. last error: %!s(<nil>)
```

**Comment by varadharajan** [comment]
I can reproduce this issue with v0.9.2 and using st1 EBS drives.

**Comment by deibyz** [comment]
Since a few months ago it is possible to resize attached EBS volumes: https://aws.amazon.com/blogs/aws/amazon-ebs-update-new-elastic-volumes-change-everything/

Some previous generation instance types don't support it though - but for the ones that do this is now possible. As volumes are always attached to the instance, they never appear as "available", but rather "in-use". Something like the patch below might do the trick?

The volume will stay in "Optimizing" for some time (potentially a long one) after the operation, so I don't know if it would be worth waiting for it to actually complete, as volume is usable during this phase.

```
$ git diff                                                                                      
diff --git a/builtin/providers/aws/resource_aws_ebs_volume.go b/builtin/providers/aws/resource_aws_ebs_volume.go
index 49d5281..1beda13 100644
--- a/builtin/providers/aws/resource_aws_ebs_volume.go
+++ b/builtin/providers/aws/resource_aws_ebs_volume.go
@@ -179,7 +179,7 @@ func resourceAWSEbsVolumeUpdate(d *schema.ResourceData, meta interface{}) error
 
                stateConf := &resource.StateChangeConf{
                        Pending:    []string{"creating", "modifying"},
-                       Target:     []string{"available"},
+                       Target:     []string{"available", "in-use"},
                        Refresh:    volumeStateRefreshFunc(conn, *result.VolumeModification.VolumeId),
                        Timeout:    5 * time.Minute,
                        Delay:      10 * time.Second,
```

**Comment by stack72** [comment]
Hi @deibyz 

You are correct - PR on it's way in with an acceptance test to prove this!

Thanks so much

Paul

**Comment by dcloud9** [comment]
Thanks for the efforts all. Which v0.9.x will the bug fix be included?

**Comment by stack72** [comment]
Hi @dcloud9 

This will be released as part of 0.9.5

@stack72
